kratos/utilities/apply_function_to_nodes_utility.cpp

```diff
@@ -0,0 +1,87 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ \.
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Riccardo Rossi
+//  Collaborator:    Vicente Mataix Ferrandiz
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "utilities/parallel_utilities.h"
+#include "utilities/apply_function_to_nodes_utility.h"
+
+namespace Kratos
+{
+
+void ApplyFunctionToNodesUtility::ApplyFunction(
+    const Variable<double>& rVariable,
+    const double t
+    )
+{    
+    // Get function
+    auto& r_function = *mpFunction;
+
+    if(!mpFunction->UseLocalSystem()) {
+        block_for_each(
+            mrNodes,r_function,
+            [&rVariable,&t](Node<3>& rNode, GenericFunctionUtility& rFunction) {
+                const double value = rFunction.CallFunction(rNode.X(), rNode.Y(), rNode.Z(), t, rNode.X0(), rNode.Y0(), rNode.Z0());
+                rNode.FastGetSolutionStepValue(rVariable) = value;
+            }
+        );
+    } else {
+        block_for_each(
+            mrNodes,r_function,
+            [&rVariable,&t](Node<3>& rNode, GenericFunctionUtility& rFunction) {
+                const double value = rFunction.RotateAndCallFunction(rNode.X(), rNode.Y(), rNode.Z(), t, rNode.X0(), rNode.Y0(), rNode.Z0());
+                rNode.FastGetSolutionStepValue(rVariable) = value;
+            }
+        );
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+std::vector<double> ApplyFunctionToNodesUtility::ReturnFunction(const double t)
+{
+    // The first node iterator
+    const auto it_node_begin = mrNodes.begin();
+
+    // The vector containing the values
+    std::vector<double> values(mrNodes.size());
+
+    // Get function
+    auto& r_function = *mpFunction;
+
+    if(!mpFunction->UseLocalSystem()) {
+        IndexPartition<std::size_t>(mrNodes.size()).for_each(r_function,
```

Comment:
u are using r_function as a thread_local_storage?? note that this is a pointer, it is not safe at all...